src/Symfony/Component/Security/Http/RememberMe/AbstractRememberMeServices.php

```diff
@@ -47,7 +47,7 @@ abstract class AbstractRememberMeServices implements RememberMeServicesInterface
     /**
      * @throws \InvalidArgumentException
      */
-    public function __construct(array $userProviders, string $secret, string $providerKey, array $options = [], LoggerInterface $logger = null)
+    public function __construct(iterable $userProviders, string $secret, string $providerKey, array $options = [], LoggerInterface $logger = null)
```

Comment:
this breaks the count() below

Reply:
I suggest changing the test to be

```php
if (is_array($userProviders) || $$userProviders instanceof \Countable) {
    $userProvidersCount = count($userProviders);
} else {
    // Fallback for non-countable iterators, to count in a way loading all providers
    $userProvidersCount = iterator_count($userProviders)
}

if (0 === $userProvidersCount) {
    // ...
}
```

This won't allow lazy-loading in the framework, because the `RewindableGenerator` injected by DI is Countable.

Reply:
>     $userProvidersCount = iterator_count($userProviders)

the iterator maybe non rewindable so we'll want to use iterator_to_array actually

Reply:
hmm, indeed. So this code:

```php
if (!\is_array($userProviders) && !$userProviders instanceof \Countable) {
    $userProviders = iterator_to_array($userProviders, false);
}

// existing conditions here.
```

Reply:
can we just drop that validation? That's what we did for similar changes IIRC. You'd get an `UnsupportedUserException` from `getUserProvider()` below. And maybe keep it for `array` and `\countable`. 
Note that within the framework this is validated at compile time (RememberMeFactory).

Reply:
Better avoid any behavior change here. Suggestion applied